File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/beecrowd/2929/Menor_da_Pilha.java
import java.util.*;

class Pilha{
    private int elementos[];
    private int topo;
    private int tamanho;

    public Pilha(int tamanho){
        this.tamanho = tamanho;
        elementos = new int[tamanho];
        topo = -1;
    }

    public boolean isVazia(){
        return topo == -1;
    }

    public boolean isFull(){
        return topo == tamanho - 1;
    }

    public void Empilhar(int elemento){
        if(isFull()){
            System.out.println("Pilha cheia! Não é possível empilhar.");
        } else{
            elementos[++topo] = elemento;
        }
    }

    public int Desempilhar(){
        int result = -1;

        if(isVazia()){
            System.out.println("Pilha vazia! Não é possível desempilhar.");
        } else{
            result = elementos[topo];
            topo--;
        }

        return result;
    }

    public int VerificarMenor(){
        int result = -1;

        if(isVazia()){
            System.out.println("A pilha está vazia!");
        } else{
            result = elementos[0];

            for(int i = 1; i <= topo; i++){
                if(elementos[i] < result){
                    result = elementos[i];
                }
            }
        }

        return result;
    }
}

public class Menor_da_Pilha{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        Pilha pilha = new Pilha(N);

        for(int i = 0; i < N; i++){
            String line = sc.next();

            if(line.equals("PUSH")){
                int V = sc.nextInt();
                pilha.Empilhar(V);
            } else if(line.equals("POP")){
                pilha.Desempilhar();
            } else if(line.equals("MIN")){
                int menor = pilha.VerificarMenor();
                if(menor != -1){
                    System.out.println(menor);
                }
            }
        }

        sc.close();
    }
}

File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/beecrowd/2929/Menor_da_Pilha.c
#include <stdio.h>
#include <string.h>

typedef struct{
    int elementos[1000];
    int topo;
    int tamanho;
} Pilha;

void iniciarPilha(Pilha *p, int tamanho){
    p->topo = -1;
    p->tamanho = tamanho;
}

int isVazia(Pilha *p){
    return p->topo == -1;
}

int isCheia(Pilha *p){
    return p->topo == p->tamanho - 1;
}

void empilhar(Pilha *p, int elemento){
    if(isCheia(p)){
        printf("Pilha cheia! Não é possível empilhar.\n");
    } else{
        p->topo++;
        p->elementos[p->topo] = elemento;
    }
}

int desempilhar(Pilha *p){
    if(isVazia(p)){
        printf("Pilha vazia! Não é possível desempilhar.\n");
        return -1;
    } else{
        int result = p->elementos[p->topo];
        p->topo--;
        return result;
    }
}

int verificarMenor(Pilha *p){
    if(isVazia(p)){
        printf("A pilha está vazia!\n");
        return -1;
    } else{
        int menor = p->elementos[0];
        for(int i = 1; i <= p->topo; i++){
            if(p->elementos[i] < menor){
                menor = p->elementos[i];
            }
        }
        return menor;
    }
}

int main(){
    int N;
    Pilha pilha;

    scanf("%d", &N);
    iniciarPilha(&pilha, N);

    for(int i = 0; i < N; i++){
        char comando[10];
        scanf("%s", comando);

        if(strcmp(comando, "PUSH") == 0){
            int valor;
            scanf("%d", &valor);
            empilhar(&pilha, valor);
        } else if(strcmp(comando, "POP") == 0){
            desempilhar(&pilha);
        } else if(strcmp(comando, "MIN") == 0){
            int menor = verificarMenor(&pilha);
            if (menor != -1){
                printf("%d\n", menor);
            }
        }
    }

    return 0;
}